rasa/shared/core/events.py

```diff
@@ -436,11 +499,14 @@ def as_sub_state(self) -> Dict[Text, Union[None, Text, List[Optional[Text]]]]:
         out = {}
         # During training we expect either intent_name or text to be set.
         # During prediction both will be set.
-        if self.intent_name:
-            out[INTENT] = self.intent_name
-        if self.text:
+        if self.text and (
```

Comment:
@wochinge  I think we really need to separate the featurization from the events, this whole substate thing isn't really a concern of this class, evident in the need to carry the e2e prediction state into this class in `use_text_for_featurization`

Reply:
`UserUttered` featurization is context dependent, so it should know how it should be featurized
the way rasa core is currently organized is that the tracker is the only one that contains the info about conversation so far, and the tracker dialogue history is represented as events.

Reply:
I am not talking about this being context-dependent, I am just saying that this isn't the right place where this logic should sit

Reply:
I disagree, how two be featurized is the property of `UserUttered` due to differences between train and test times

Reply:
I agree that the `as_sub_state` would be to be a part of a certain featurizer. In my opinion this is similar to the discussion about having the representation of an event in a certain format within the `Event` class or within a different class (e.g. `YAMLStoryReader`). It works as long as there is only one way to featuring things, but gets nasty as soon as we have multiple ways of featurizing things.